CAPV controller panics if underlying VSphereVMs are deleted

/kind bug


After doing some manual deletions of VMs/ removing some finalizers on my cluster objects, i saw capv controller manager panicking...

```
Hi folks, I just (after doing some manual things in vsphere to delete some machiens) ,saw a panic ….
I0830 08:34:54.527424       1 reflector.go:257] Listing and watching *v1.Secret from pkg/mod/k8s.io/client-go@v0.26.1/tools/cache/reflector.go:169                                                                                                    
I0830 08:34:54.526331       1 controller.go:118] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="vspherevm" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="VSphereVM
" VSphereVM="default/windows-cluster-dt6bx-89h92" namespace="default" name="windows-cluster-dt6bx-89h92" reconcileID=e34e3a75-4b2f-4948-a143-1ad972e433ee                                                                                             
I0830 08:34:54.526551       1 controller.go:118] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="vspherevm" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="VSphereVM
" VSphereVM="default/windows-cluster-dt6bx-vw75j" namespace="default" name="windows-cluster-dt6bx-vw75j" reconcileID=7e9a7012-597c-48d3-9046-dcb8c91e08cd                                                                                             
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                                                                                                   
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                       
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1b6fbf6]                                                                                                                                                                              
                                                                                                                                                                                                                                                      
goroutine 759 [running]:                                                                                                                                                                                                                              
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()                                                                                                                                                                
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:119 +0x1fa                                                                                                                                           
panic({0x1e89580, 0x38a03b0})                                                                                                                                                                                                                         
        /usr/local/go/src/runtime/panic.go:884 +0x212                                                                                                                                                                                                 
sigs.k8s.io/cluster-api-provider-vsphere/controllers.vmReconciler.retrieveVcenterSession({0xc000a0ba40?, {0x25e9a88?, 0x3911cb8?}}, {0x25f9a38, 0xc0009f4300}, 0xc000448380)                                                                          
        /workspace/controllers/vspherevm_controller.go:559 +0x296                                                                                                                                                                                     
sigs.k8s.io/cluster-api-provider-vsphere/controllers.vmReconciler.Reconcile({0xc000a0ba40?, {0x25e9a88?, 0x3911cb8?}}, {0x25f9a38, 0xc0009f4300}, {{{0xc000a26d26, 0x7}, {0xc0006bc620, 0x1b}}})                                                      
        /workspace/controllers/vspherevm_controller.go:186 +0x34e                                                                                                                                                                                     
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x25f9a38?, {0x25f9a38?, 0xc0009f4300?}, {{{0xc000a26d26?, 0x1d7c900?}, {0xc0006bc620?, 0x0?}}})                                                                       
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:122 +0xc8                                                                                                                                            
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0005eaaa0, {0x25f9990, 0xc0009be200}, {0x1fb2c40?, 0xc0003f0b40?})                                                                                           
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:323 +0x3a5                                                                                                                                           
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0005eaaa0, {0x25f9990, 0xc0009be200})                                                                                                                     
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:274 +0x1d9                                                                                                                                           
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()                                                                                                                                                                  
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:235 +0x85                                                                                                                                            
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2                                                                                                                                                           
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.5/pkg/internal/controller/controller.go:231 +0x333
```
will add more details shortly...


**What steps did you take and what happened:**

Deleted some infra manually

**What did you expect to happen:**

CAPV would gracefully tolerate changes to missing objects and update its caches however it needs to, rather than panicking b/c things are gone.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster-api-provider-vsphere version: v1.7.0_vmware.0                                                                                                                                            
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

--- a/controllers/vspherevm_controller.go
+++ b/controllers/vspherevm_controller.go
@@ -28,6 +28,7 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
 	"k8s.io/utils/pointer"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/controllers/remote"
@@ -569,6 +570,9 @@ func (r vmReconciler) retrieveVcenterSession(ctx goctx.Context, vsphereVM *infra
 			params)
 	}
 
+	if cluster.Spec.InfrastructureRef == nil {
+		return nil, errors.Errorf("cannot retrieve vCenter session for cluster %s: .spec.infrastructureRef is nil", klog.KObj(cluster))
+	}
 	key := ctrlclient.ObjectKey{
 		Namespace: cluster.Namespace,
 		Name:      cluster.Spec.InfrastructureRef.Name,

--- a/controllers/vspherevm_controller_test.go
+++ b/controllers/vspherevm_controller_test.go
@@ -420,18 +420,6 @@ func TestRetrievingVCenterCredentialsFromCluster(t *testing.T) {
 		},
 	}
 
-	cluster := &clusterv1.Cluster{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "valid-cluster",
-			Namespace: "test",
-		},
-		Spec: clusterv1.ClusterSpec{
-			InfrastructureRef: &corev1.ObjectReference{
-				Name: vsphereCluster.Name,
-			},
-		},
-	}
-
 	machine := &clusterv1.Machine{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "foo",
@@ -441,8 +429,6 @@ func TestRetrievingVCenterCredentialsFromCluster(t *testing.T) {
 			},
 		},
 	}
-
-	initObjs := createMachineOwnerHierarchy(machine)
 
 	vsphereMachine := &infrav1.VSphereMachine{
 		ObjectMeta: metav1.ObjectMeta{
@@ -480,26 +466,71 @@ func TestRetrievingVCenterCredentialsFromCluster(t *testing.T) {
 		Status: infrav1.VSphereVMStatus{},
 	}
 
-	initObjs = append(initObjs, secret, vsphereVM, vsphereMachine, machine, cluster, vsphereCluster)
-	controllerMgrContext := fake.NewControllerManagerContext(initObjs...)
+	t.Run("Retrieve credentials from cluster", func(t *testing.T) {
+		cluster := &clusterv1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "valid-cluster",
+				Namespace: "test",
+			},
+			Spec: clusterv1.ClusterSpec{
+				InfrastructureRef: &corev1.ObjectReference{
+					Name: vsphereCluster.Name,
+				},
+			},
+		}
 
-	controllerContext := &context.ControllerContext{
-		ControllerManagerContext: controllerMgrContext,
-		Recorder:                 record.New(apirecord.NewFakeRecorder(100)),
-		Logger:                   log.Log,
-	}
-	r := vmReconciler{ControllerContext: controllerContext}
+		initObjs := createMachineOwnerHierarchy(machine)
+		initObjs = append(initObjs, secret, vsphereVM, vsphereMachine, machine, cluster, vsphereCluster)
+		controllerMgrContext := fake.NewControllerManagerContext(initObjs...)
 
-	_, err = r.Reconcile(goctx.Background(), ctrl.Request{NamespacedName: util.ObjectKey(vsphereVM)})
-	g := NewWithT(t)
-	g.Expect(err).NotTo(HaveOccurred())
+		controllerContext := &context.ControllerContext{
+			ControllerManagerContext: controllerMgrContext,
+			Recorder:                 record.New(apirecord.NewFakeRecorder(100)),
+			Logger:                   log.Log,
+		}
+		r := vmReconciler{ControllerContext: controllerContext}
 
-	vm := &infrav1.VSphereVM{}
-	vmKey := util.ObjectKey(vsphereVM)
-	g.Expect(r.Client.Get(goctx.Background(), vmKey, vm)).NotTo(HaveOccurred())
-	g.Expect(conditions.Has(vm, infrav1.VCenterAvailableCondition)).To(BeTrue())
-	vCenterCondition := conditions.Get(vm, infrav1.VCenterAvailableCondition)
-	g.Expect(vCenterCondition.Status).To(Equal(corev1.ConditionTrue))
+		_, err = r.Reconcile(goctx.Background(), ctrl.Request{NamespacedName: util.ObjectKey(vsphereVM)})
+		g := NewWithT(t)
+		g.Expect(err).NotTo(HaveOccurred())
+
+		vm := &infrav1.VSphereVM{}
+		vmKey := util.ObjectKey(vsphereVM)
+		g.Expect(r.Client.Get(goctx.Background(), vmKey, vm)).NotTo(HaveOccurred())
+		g.Expect(conditions.Has(vm, infrav1.VCenterAvailableCondition)).To(BeTrue())
+		vCenterCondition := conditions.Get(vm, infrav1.VCenterAvailableCondition)
+		g.Expect(vCenterCondition.Status).To(Equal(corev1.ConditionTrue))
+	},
+	)
+
+	t.Run("Error if cluster infrastructureRef is nil", func(t *testing.T) {
+		cluster := &clusterv1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "valid-cluster",
+				Namespace: "test",
+			},
+
+			// InfrastructureRef is nil so we should get an error.
+			Spec: clusterv1.ClusterSpec{
+				InfrastructureRef: nil,
+			},
+		}
+		initObjs := createMachineOwnerHierarchy(machine)
+		initObjs = append(initObjs, secret, vsphereVM, vsphereMachine, machine, cluster, vsphereCluster)
+		controllerMgrContext := fake.NewControllerManagerContext(initObjs...)
+
+		controllerContext := &context.ControllerContext{
+			ControllerManagerContext: controllerMgrContext,
+			Recorder:                 record.New(apirecord.NewFakeRecorder(100)),
+			Logger:                   log.Log,
+		}
+		r := vmReconciler{ControllerContext: controllerContext}
+
+		_, err = r.Reconcile(goctx.Background(), ctrl.Request{NamespacedName: util.ObjectKey(vsphereVM)})
+		g := NewWithT(t)
+		g.Expect(err).To(HaveOccurred())
+	},
+	)
 }
 
 func Test_reconcile(t *testing.T) {

